src/backend/distributed/metadata/pg_get_object_address_12_13_14.c

```diff
@@ -0,0 +1,536 @@
+		{
+			objnode = (Node *) list_make2(name, linitial(args));
+			break;
+		}
+
+		case OBJECT_USER_MAPPING:
+		{
+			objnode = (Node *) list_make2(linitial(name), linitial(args));
+			break;
+		}
+
+		case OBJECT_DEFACL:
+		{
+			objnode = (Node *) lcons(linitial(args), name);
+			break;
+		}
+
+		case OBJECT_AMOP:
+		case OBJECT_AMPROC:
+		{
+			objnode = (Node *) list_make2(name, args);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_ROUTINE:
+		case OBJECT_AGGREGATE:
+		case OBJECT_OPERATOR:
+		{
+			ObjectWithArgs *owa = makeNode(ObjectWithArgs);
+
+			owa->objname = name;
+			owa->objargs = args;
+			objnode = (Node *) owa;
+			break;
+		}
+
+		case OBJECT_LARGEOBJECT:
+		{
+			/* already handled above */
+			break;
+		}
+
+			/* no default, to let compiler warn about missing case */
+	}
+
+	if (objnode == NULL)
+	{
+		elog(ERROR, "unrecognized object type: %d", type);
+	}
+
+	ObjectAddress addr = get_object_address(type, objnode,
+											&relation, AccessShareLock, false);
+
+	/* CITUS CODE BEGIN */
+	ErrorIfCurrentUserCanNotDistributeObject(type, &addr, objnode, &relation);
+
+	/* CITUS CODE END */
+
+	/* We don't need the relcache entry, thank you very much */
+	if (relation)
+	{
+		relation_close(relation, AccessShareLock);
+	}
+
+	/* CITUS CODE BEGIN */
+	return addr;
+
+	/* CITUS CODE END */
+}
+
+
+/*
+ * ErrorIfCurrentUserCanNotDistributeObject checks whether current user can
+ * distribute object, if not errors out.
+ */
+static void
+ErrorIfCurrentUserCanNotDistributeObject(ObjectType type, ObjectAddress *addr,
+										 Node *node, Relation *relation)
+{
+	Oid userId = GetUserId();
+	AclMode aclMaskResult = 0;
+	bool passAclCheck = false;
+	Oid idToCheck = InvalidOid;
+
+	/* Since we don't handle sequences like object, add it separately */
+	if (!(SupportedDependencyByCitus(addr) || type == OBJECT_SEQUENCE))
+	{
+		ereport(ERROR, (errmsg("Object type %d can not be distributed by Citus", type)));
+	}
+
+	switch (type)
+	{
+		case OBJECT_ACCESS_METHOD:
+		{
+			if (IsObjectAddressOwnedByExtension(addr, NULL))
+			{
+				ereport(ERROR, (errmsg("Current user does not have required "
+									   "access privileges on access method %d with type %d",
+									   addr->objectId, type)));
+			}
+			passAclCheck = true;
+			break;
+		}
+
+		case OBJECT_SCHEMA:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_namespace_aclmask(idToCheck, userId, ACL_USAGE,
+												 ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_AGGREGATE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_proc_aclmask(idToCheck, userId, ACL_EXECUTE,
+											ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_DATABASE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_database_aclmask(idToCheck, userId, ACL_CONNECT,
+												ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_ROLE:
+		{
+			/* Support only extension owner role with community */
```

Comment:
minor: we might avoid extra paranthesis: `addr->objectId != CitusExtensionOwner()`?